substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java

```diff
@@ -0,0 +1,185 @@
+import java.util.stream.Collectors;
+
+/**
+ * This feature:
+ * <ul>
+ *      <li> synthesizes the runtime boot module layer </li>
+ *      <li> ensures that fields/methods from the {@link ClassLoader} class are reachable in order to
+ *      make native methods of the {@link Module} class work </li>
+ * </ul>
+ * <p>
+ * This feature synthesizes the runtime boot module layer by using type reachability information.
+ * If a type is reachable, its module is also reachable and therefore should be included in the
+ * runtime boot module layer.
+ * </p>
+ * <p>
+ * The configuration for the runtime boot module layer is resolved using the module reachability
+ * data provided to us by the analysis as resolve roots.
+ * </p>
+ * <p>
+ * We are purposefully avoiding public API for module layer creation, such as
+ * {@link ModuleLayer#defineModulesWithOneLoader(Configuration, ClassLoader)}, because as a side
+ * effect this will create a new class loader. Instead, we use a private constructor to construct
+ * the {@link ModuleLayer} instance, which we then patch using the module reachability data
+ * provided to us by the analysis.
+ * </p>
+ * <p>
+ * Because the result of this feature is dependant on the analysis results, and because this feature
+ * will add reachable object(s) as it's result, it is necessary to perform the logic during the
+ * analysis, for lack of a better option (even though only the last analysis cycle is sufficient,
+ * but that cannot be known in advance).
+ * </p>
+ */
+@AutomaticFeature
+@Platforms(Platform.HOSTED_ONLY.class)
+public final class ModuleLayerFeature implements Feature {
+
+    private Field moduleNameToModuleField;
+    private Field moduleParentsField;
+    private Constructor<ModuleLayer> moduleLayerConstructor;
+
+    @Override
+    public boolean isInConfiguration(IsInConfigurationAccess access) {
+        return new JDK11OrLater().getAsBoolean();
+    }
+
+    @Override
+    public void afterRegistration(AfterRegistrationAccess access) {
+        ImageSingletons.add(BootModuleLayerSupport.class, new BootModuleLayerSupport());
+        moduleNameToModuleField = ReflectionUtil.lookupField(ModuleLayer.class, "nameToModule");
+        moduleParentsField = ReflectionUtil.lookupField(ModuleLayer.class, "parents");
+        moduleLayerConstructor = ReflectionUtil.lookupConstructor(ModuleLayer.class, Configuration.class, List.class, Function.class);
+    }
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
+        access.registerReachabilityHandler(
+                a -> a.registerAsUnsafeAccessed(ReflectionUtil.lookupField(ClassLoader.class, "classLoaderValueMap")),
+                ReflectionUtil.lookupMethod(ClassLoader.class, "trySetObjectField", String.class, Object.class)
+        );
+        RuntimeReflection.register(ReflectionUtil.lookupField(ClassLoader.class, "classLoaderValueMap"));
+    }
+
+    @Override
+    public void duringAnalysis(DuringAnalysisAccess access) {
```

Comment:
IMO, this should be
```diff
@@ -115,11 +115,11 @@
     }
 
     @Override
-    public void duringAnalysis(DuringAnalysisAccess access) {
-        FeatureImpl.DuringAnalysisAccessImpl accessImpl = (FeatureImpl.DuringAnalysisAccessImpl) access;
+    public void afterAnalysis(AfterAnalysisAccess access) {
+        FeatureImpl.AfterAnalysisAccessImpl accessImpl = (FeatureImpl.AfterAnalysisAccessImpl) access;
         AnalysisUniverse universe = accessImpl.getUniverse();
 
         Map<String, Module> reachableModules = universe.getTypes()
                 .stream()
                 .filter(t -> t.isReachable() && !t.isArray())
                 .map(t -> t.getJavaClass().getModule())
```

Reply:
Using `@UnknownObjectField` with `ModuleLayer` as a provided type does indeed allow for this. The only question that remains is what should be manually marked as reachable? I assume only the `ModuleLayer` and `Configuration` types, although if my reasoning is correct, they are always reachable anyway.